ZEP008-hot-backup.md

```diff
@@ -0,0 +1,619 @@
+
+That means an event e1 happens before e2 iff:
+- both events happened in the same process and e1 occurred before e2
+- e1 sends a message and e2 receives it
+- e1 happened before e2 transitively
+
+A cut is a subset of events in the state.
+A cut C is consistent if : $\forall e_i, e_j : e_j \in C \wedge e_i \rightarrow e_j \implies e_i \in C$.
+That means in consistent cut C, if an event exists in C then all events *happened before* it must also exist in C.
+
+
+![consistent-cut](ZEP008/consistent-cut.png)
+
+In the above image, the horizontal line depicts the time in each process. The vertical slanted arrow shows communication from one process to another. Dotted lines represent a cut. That means - we take a checkpoint where the dotted line meets the horizontal timeline. The green ones are a consistent cut. The red lines are cuts, but they are not consistent cuts. The red line is not consistent because e2 is in the cut, but e1 is not. e1 has happened before e2 because it is a "send-receive" pair.
+
+### Algorithm to find a consistent cut
+
+[BCS algorithm as proposed in this paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.149.5565&rep=rep1&type=pdf) is a way to take coordinated consistent checkpoints across multiple processes. (An easy to understand version is [here](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.78.4041&rep=rep1&type=pdf)). This is a communication induced checkpointing and relies on lazy coordination among the participating processes.
+
+Here is how the algorithm works.
+
+$C_i^{sn_i}$ is a local checkpoint of process $i$, and its checkpoint id is ${sn_i}$. When a process sends a message, it also embeds it current checkpoint id ${sn_i}$ with it.
+
+A process takes it local checkpoint when
+1. A basic checkpoint is triggered. A basic checkpoint is triggered either by an external trigger or it could be something that is triggered periodically.
+2. A forced checkpoint is triggered. A forced checkpoint happens when a process received a message from another process, and it is forced to take a checkpoint to guarantee the consistency.
+
+The algorithm is as follows:
+
+*On basic checkpoint:*  
+&emsp; $sn_i = sn_i + 1$  
+&emsp; take checkpoint $C_i^{sn_i}$
+
+*Upon the receipt of a message m:*  
+&emsp; if $sn_i < m.sn$ then  
+&emsp; &emsp; $sn_i = m.sn$  
+&emsp; &emsp; take checkpoint $C_i^{sn_i}$
+&emsp; process message $m$
+
+It has been proven that if we follow the above algorithm, a global checkpoint $C = \{ C_i^{sn_i}, C_j^{sn_k}, C_l^{sn_l}, ...\}$ is a consistent cut if all local checkpoints in $C$ has the same id. $sn_i = sn_j = sn_l = ...$.
+
+#### Correctness of high-level process
+
+The high level process explained in [Guide-level explanation](#internal-backup-process) follows implements the BCS algorithm.
+
+A record in the logstream represents an event in the system. The position of the event can be considered as the logical time when the event occurred.
```

Comment:
:question: I suppose here we can ignore consensus issues? What would happen if a leader change occurs during a checkpoint? Could it cause issues, where position X was a different record previously?